Import ProGraML protocol buffers and python builds.

ChrisCummins opened the pull request:
This imports the ProGraML protocol buffers from:

Remote:   https://github.com/ChrisCummins/ProGraML.git
Commit:   90f0309caabc5014a95c2ce438c4c5f17637c9b1

And modifies `deeplearning.ml4pl.poj104.dataset.nx2data()` to load these protocol buffers rather than pickled nx graphs.

This adds a new protobuf dependency, see `requirements.txt`.

Run ad-hoc "tests" using:
```
$ ./chris_tests.sh
=== programl/proto/protos_test.py
Loaded graph with 226 nodes

=== deeplearning/ml4pl/poj104/dataset_test.py
Loaded graph with 226 nodes and 404 edges
Loaded 2230 element vocabulary
Created unlabelled graph data Data(edge_attr=[404, 2], edge_index=[2, 404], x=[226, 2])
Created POJ-104 labelled graph data Data(edge_attr=[404, 2], edge_index=[2, 404], x=[226, 2], y=[1])
```

ChrisCummins committed:
Import ProGraML protocol buffers and python builds.

This imports the ProGraML protocol buffers from:

Remote:   https://github.com/ChrisCummins/ProGraML.git
Commit:   90f0309caabc5014a95c2ce438c4c5f17637c9b1

And commits the python code generated using:

    $ protoc programl/proto --python_out=.

ChrisCummins committed:
Don't track *.pyc files or virtualenv

ChrisCummins committed:
Update nx2graph() to read ProgramGraph protocol buffers.

Test using:

    $ ./chris_tests.sh

ChrisCummins committed:
Add a requirements.txt file.

Zacharias030 committed:
fixing the POJ104Dataset process() method and new debug config

Zacharias030 committed:
added warning for breaking code in nx2data

Zacharias030 committed:
fix config "json" parsing and hotfix for poj104 labels between [1,104] instead of [0,104)

ChrisCummins commented on deeplearning/ml4pl/models/ggnn/run.py:
```diff
@@ -353,7 +353,9 @@ def data2input(self, batch):
-                "labels": batch.y,
+                # TODO this is a hotfix and not required after re-collating the data_train.pb
+                # of the rebuttal-classifyapp-dataset
+                "labels": batch.y - 1,
```
Is this duplicated with -1 offset added to `nx2data()` below?

Zacharias030 replied:
yes, but it was working if you still had the old preprocessing cached. I un-hotfixed this line below and reprocessed my data. should all work now (since yesterday evening).

ChrisCummins replied:
Great! If the results are looking good, let's merge and I'll submit a new PR for devmap

Zacharias030 committed:
un-hotfix after redoing model preprocessing

ChrisCummins committed:
Add a param to nx2data() to use ablated vocab.

Zacharias030 committed:
Merge pull request #2 from Zacharias030/chris/ablation_vocab

Add a param to nx2data() to use ablated vocab

ChrisCummins committed:
Add support for converting to CDFG format at load time.

This adds a cdfg flag to the dataset load() operation that converts a
ProGraML graph to CDFG format after loading. This also adds a CDFG
vocabulary file.

Care must be take to ensure that the cached files don't conflict
between the CDFG/non-CDFG versions.

ChrisCummins committed:
Add support for loading the new devmap dataset.

This supports the new proto-based dataset, and also CDFG format.

Zacharias030 commented on deeplearning/ml4pl/poj104/dataset.py:
```diff
@@ -24,56 +30,84 @@
-        try:
-            data = pickle.load(f)
-        except EOFError as e:
-            print(f"Failing on {str(file)}")
-            raise e
-    return data
+        graph.ParseFromString(f.read())
+    return graph
+
+
+def load_vocabulary(path: Path):
+    """Read the vocabulary file used in the dataflow experiments."""
+    vocab = {}
+    with open(path) as f:
+        vocab_file = csv.reader(f.readlines(), delimiter="\t")
+        for i, row in enumerate(vocab_file, start=-1):
+            if i == -1:  # Skip the header.
+                continue
+            (_, _, _, text) = row
+            vocab[text] = i
+
+    return vocab
 
 
-def nx2data(nx_graph, class_label=None, ignore_profile_info=True):
-    r"""Converts a :obj:`networkx.Graph` or :obj:`networkx.DiGraph` to a
+class AblationVocab(enum.IntEnum):
+    # No ablation - use the full vocabulary (default).
+    NONE = 0
+    # Ignore the vocabulary - every node has an x value of 0.
+    NO_VOCAB = 1
+    # Use a 3-element vocabulary based on the node type:
+    #    0 - Instruction node
+    #    1 - Variable node
+    #    2 - Constant node
+    NODE_TYPE_ONLY = 2
```
Note to self:
the ablation state is part of the cached datasets, but the filenames in the cache don't represent that. This could lead to unwanted mixups.

ChrisCummins replied:
Yes, good point!

Zacharias030 reviewed (commented):
Should we better mark the ablation state in the cached datasets as well?

ChrisCummins committed:
Merge pull request #3 from Zacharias030/chris/cdfg

Add support for converting to CDFG format at load time

ChrisCummins committed:
Prevent filename conflicts with different combos of vocab options.

Zacharias030 committed:
Merge pull request #4 from Zacharias030/chris/vocab_opts

Prevent filename conflicts with different combos of vocab options

Zacharias030 commented:
I think we are ready! Very nice changes :)

I'll merge and then test by running experiments.
